R/filters.R

```diff
@@ -20,38 +20,57 @@
 #' repFilter(immdata, "by.meta", list(Lane = c("A", "B"), Age = ">=15"))
 #' @export repFilter
 repFilter <- function(.data, .method, .query, ...) {
+  if (!is.list(.data)) {
+    stop(paste0(
+      "Input data is not a list; ",
+      "please pass Immunarch object with data and metadata as input."
+    ))
+  }
+  if (!("meta" %in% names(.data))) {
+    stop(paste0(
+      "Input data doesn't contain meta; ",
+      "please pass Immunarch object with data and metadata as input."
+    ))
+  }
+
   switch(tolower(.method),
     by.meta = filter_by_meta(.data, .query),
-    stop("You entered the wrong method! Please, try again.")
+    stop("You entered the wrong method! Supported methods are: \"by.meta\".")
   )
 }
 
 filter_by_meta <- function(.data, .query) {
-  original_classes <- class(.data)
   names <- names(.query)
 
   filtered_meta <- .data$meta
   for (i in seq_along(names)) {
     name <- names[i]
-    column_query <- unlist(.query[name])
-    filtered_meta <-
-      if (length(column_query) > 1) {
-        filter(filtered_meta, get(name) %in% column_query)
-      } else if (startsWith(column_query, ">=")) {
-        filter(filtered_meta, get(name) >= as.numeric(substring(column_query, 3)))
-      } else if (startsWith(column_query, "<=")) {
-        filter(filtered_meta, get(name) <= as.numeric(substring(column_query, 3)))
-      } else if (startsWith(column_query, ">")) {
-        filter(filtered_meta, get(name) > as.numeric(substring(column_query, 2)))
-      } else if (startsWith(column_query, "<")) {
-        filter(filtered_meta, get(name) < as.numeric(substring(column_query, 2)))
-      } else {
-        filter(filtered_meta, get(name) == column_query)
+    if (!(name %in% names(.data$meta))) {
+      stop(paste0("Column \"", name, "\" not found in metadata."))
+    }
+    column_query <- .query[[name]]
+
+    if (nrow(filtered_meta) > 0) {
+      filtered_meta <-
+        if (length(column_query) > 1) {
+          filter(filtered_meta, get(name) %in% column_query)
+        } else if (startsWith(column_query, ">=")) {
+          filter(filtered_meta, get(name) >= as_numeric_or_fail(substring(column_query, 3)))
+        } else if (startsWith(column_query, "<=")) {
+          filter(filtered_meta, get(name) <= as_numeric_or_fail(substring(column_query, 3)))
+        } else if (startsWith(column_query, ">")) {
+          filter(filtered_meta, get(name) > as_numeric_or_fail(substring(column_query, 2)))
+        } else if (startsWith(column_query, "<")) {
+          filter(filtered_meta, get(name) < as_numeric_or_fail(substring(column_query, 2)))
+        } else {
+          filter(filtered_meta, get(name) == column_query)
+        }
+      if (nrow(filtered_meta) == 0) {
+        warning(paste0("Filter by column \"", name, "\" removed all remaining samples!"))
       }
+    }
   }
   filtered_data <- .data$data[filtered_meta$Sample]
```

Comment:
Did you check that this works with `data.table` and database instance?

Reply:
I wrote it in comments for previous PR: `.data$data` is always a list, and `.data$meta` is always tibble dataframe. They can't be anything else: https://github.com/immunomind/immunarch/blob/dev/R/io.R#L106